### PR TITLE
automation: add AI-powered issue quality analysis workflow

### PR DESCRIPTION
**Pull request by modichika**
**Description of your changes:**
An AI-Powered issue analyzer that analyzes GitHub issues as soon as it opens. It uses 'gpt-4o-mini' via GitHub Models which is free, fast, and provides clear reasoning devided under: scope, context and guidance, complexity, overall issue quality verdict.

**Addressing the problem.**
Repositories that are high-usage to the tech community, its maintainers often are under the gun to analyse issues. That's a pain point I felt while finding the right issue that i need to work on for kubeflow/pipelines. 

**Why this is needed?**
Analysing issues is one kind of job in itself when maintainers are busy and too under the gun.

**The solutions:**
- A GitHub Action that triggers immediately upon issue creation.
- Wrote a ai_analyzer.yml file in .github/workflows/ai_analyzer.yml, utilize the built-in GITHUB_TOKEN for security and simplicity.
- Included a script folder suggested by @alyssacgoins so that the main .yml file remains clean and we have a strict prompt for llm.
- I included the comment one where ai leaves a comment after labelling of why it selected those labels, this provides transparency and allows maintainers to verify the AI's logic

**Checklist:**
- [x] You have [signed off your commits]
- [x] The title for your pull request (PR) should follow our title convention.

### CODE REVIEW

**Comment by droctothorpe** [comment]
@alyssacgoins Hey does this align with your grooming KEP? I want to make sure that we're not duplicating efforts.

**Comment by modichika** [comment]
Hey @droctothorpe the already labeled good first issue in official kubeflow/pipelines i ran my updated version it labeled only one as good first issue and others as not as they required knowledge of codebase, can you test on them as well?

**Comment by alyssacgoins** [comment]
> @alyssacgoins Hey does this align with your grooming KEP? I want to make sure that we're not duplicating efforts.

hey @droctothorpe this does overlap with my [KEP](https://github.com/kubeflow/pipelines/pull/13293) - but I really like the work here. I see you've added some feedback on my KEP (thank you! will address next week when I return from PTO) - @modichika would you mind taking a look at my KEP as well? 
My thinking is that we collaborate together on the KEP, get that merged, and I'll create a secondary PR that addresses anything out of scope for this PR.

**Comment by modichika** [comment]
Hey @alyssacgoins thank you for the review. 
Yeah I would like to collaborate with you on the KEP.
I have looked into the KEP but I'll take a deeper look.

File: .github/workflows/ai_labeler.yml [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+          GH_TOKEN: ${{ secrets.GITHUB_TOKEN }}
+          TITLE: ${{ github.event.issue.title }}
+          BODY: ${{ github.event.issue.body }}
+        run: |
+          # 1. Install and run in the same step
+          gh extension install github/gh-models
+
+
+          # 2. Strict prompt to get ONLY good first issue label
+          RESULT=$(gh models run openai/gpt-4o-mini "You are a maintainer of the repo.
```
Comment (alyssacgoins) [inline .github/workflows/ai_labeler.yml]: Do we need to worry about the long-term availability of this model for our use case? It's free for now - what is the plan for if this model becomes paid?

File: .github/workflows/ai_labeler.yml [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+          TITLE: ${{ github.event.issue.title }}
+          BODY: ${{ github.event.issue.body }}
+        run: |
+          # 1. Install and run in the same step
+          gh extension install github/gh-models
+
+
+          # 2. Strict prompt to get ONLY good first issue label
+          RESULT=$(gh models run openai/gpt-4o-mini "You are a maintainer of the repo. 
+          Analyze if this issue is a good first issue based on these rules:
```
Comment (alyssacgoins) [inline .github/workflows/ai_labeler.yml]: Did you compare `openai/gpt-4o-mini` against any other similar options?

File: .github/workflows/ai_labeler.yml [hunk 3]
```diff
@@ -0,0 +1,70 @@
+name: AI Issue Labeler
+on:
+  issues:
+    types: [opened]
```
Comment (alyssacgoins) [inline .github/workflows/ai_labeler.yml]: I think this should also run when an existing issue is modified

**Review by alyssacgoins — CHANGES_REQUESTED** [review]
Have you tested this on your KFP fork? It would be useful to see some examples of this workflow in action

File: .github/workflows/ai_labeler.yml (hunk 1 above)
Comment (modichika) [reply .github/workflows/ai_labeler.yml]: Hey so i tested it and once it did fail with the rate limit error but the thing is before the issue reaches the llm we can make the issue template mandatory?

File: .github/workflows/ai_labeler.yml (hunk 3 above)
Comment (modichika) [reply .github/workflows/ai_labeler.yml]: updated this @alyssacgoins to address the modified ones

**Comment by modichika** [comment]
yeah i did ran tests for it, it passed for the good first issues here's it : https://github.com/modichika/pipelines/actions/runs/25742713067/job/75597340390
@alyssacgoins you can look it labelled good first issue

But needs to address with the modified ones as well and with the fallbacks you mentioned.

File: .github/workflows/ai_labeler.yml (hunk 2 above)
Comment (modichika) [reply .github/workflows/ai_labeler.yml]: I haven't until now but i will compare it as well

**Comment by alyssacgoins** [comment]
Hey @modichika,
@mprahl and I were thinking it would be useful to introduce a human component to your workflow, something like the following - 
1. As in your current workflow, when an issue is created or modified, the workflow kicks off and sends a request to `gpt-4o-mini` through `GH models`
2. Instead of prompting the model to identify if the issue is a good first issue, prompt the model to analyze the issue for general scope, context guidance and complexity (the same parameters you currently query) and respond with a general analysis for issue quality. You can also give the model several pre-approved issues for context on what a good issue looks like.
3. Prompt the model to use both the issue header and body to identify the section of the codebase targeted by the issue (issue header will often contain the section name outright). 
4. The script would then tag the owners and approvers for that section of the codebase (found in the OWNERS.md file for a subdirectory), paste in the feedback from `gpt-4o-mini` and ask those users to use their judgement and the AI analysis to decide if the issue should be labeled `/ready`.

What are your thoughts?

CC @droctothorpe

**Comment by modichika** [comment]
Hey @alyssacgoins it's amazing let's implement this.
I'll implement pt. 2.
I need a little bit more context towards pt. 3 and 4, it would be great if it's clear exactly what's needed.
what do you think?

**Comment by modichika** [comment]
@alyssacgoins made a new commit in the code to address pt. 2 here's a test output: https://github.com/modichika/pipelines/issues/58

Also lmk about the pre-approved issues which ones to give exactly.

I noticed one thing tho without the issue form that users fills before opening the issue, the llm might hallucinate and point out small things that might have been covered through the form for context and do not look at the complexity of the issue.

So maybe including the form will be good, what do you think?

File: .github/workflows/ai_labeler.yml [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+    runs-on: ubuntu-latest
+    permissions:
+      issues: write
+      contents: read
+      models: read
+    steps:
+      - name: Checkout
+        uses: actions/checkout@v4
+
+      - name: Get AI Label
```
Comment (alyssacgoins) [inline .github/workflows/ai_labeler.yml]:
```suggestion
      - name: Execute AI Issue Triage
```
(or something similar)

File: .github/workflows/ai_analyzer.yml [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+      - name: Checkout
+        uses: actions/checkout@v4
+
+      - name: Get AI Label
+        id: ai_logic
+        env:
+          GH_TOKEN: ${{ secrets.GITHUB_TOKEN }}
+          TITLE: ${{ github.event.issue.title }}
+          BODY: ${{ github.event.issue.body }}
+        run: |
```
Comment (alyssacgoins) [inline .github/workflows/ai_analyzer.yml:26]: Consider placing this script in an external .sh file under `.github/scripts` rather than inline

**Comment by alyssacgoins** [comment]
@modichika Great work! I added some minor comments, but this is exactly what we were thinking. I agree that it would be good context to include the issue template in the prompt. 
Here are some good examples of issues for each section of the codebase:

- **backend**: https://github.com/kubeflow/pipelines/issues/13314 
- **bug**: https://github.com/kubeflow/pipelines/issues/13180
- **frontend**: https://github.com/kubeflow/pipelines/issues/13108
- **sdk**: https://github.com/kubeflow/pipelines/issues/12865

**Comment by modichika** [comment]
Hey @alyssacgoins added the pre-approved issues in the prompt including some tests and my observations here:

1. https://github.com/modichika/pipelines/issues/59  the difference between 1st llm's output and 2nd one is that in the 2nd one the template was filled accurately but the 1st one lacked two sections and you can see the llm's output improved.

2. https://github.com/modichika/pipelines/issues/60  in this issue testing, the token was so much that the llm couldn't handle so we need to think around this edge case.

Lmk your thoughts!

**Comment by alyssacgoins** [comment]
> Hey @alyssacgoins added the pre-approved issues in the prompt including some tests and my observations here:
> 
> 1. [[sdk]  [bug] [set_accelerator_limit] rejects valid accelerator counts (only allows 0, 1, 2, 4, 8, 16) modichika/pipelines#59](https://github.com/modichika/pipelines/issues/59)  the difference between 1st llm's output and 2nd one is that in the 2nd one the template was filled accurately but the 1st one lacked two sections and you can see the llm's output improved.
> 2. [fix: E2E test flakiness on K8s v1.34.0 — root cause analysis modichika/pipelines#60](https://github.com/modichika/pipelines/issues/60)  in this issue testing, the token was so much that the llm couldn't handle so we need to think around this edge case.
> 
> Lmk your thoughts!

@modichika Glad to see such a noticeable change from adding the issue context - this is exactly what I was hoping for. Can you do some research and identify a couple slightly more powerful models offered through `GH models ` to test out on the second larger issue? We don't want people to be able deterred from creating detailed issues, so finding a powerful enough model is key here. 
Some supported models you can try:
- [OpenAI o4-mini](https://github.com/marketplace/models/azure-openai/o4-mini/playground) - while the name is quite similar to gpt-4o-mini, `o4` is specifically a reasoning model, and contains an expanded token window
- [Mistral Codestral 25.01](https://github.com/marketplace/models/azureml-mistral/Codestral-2501/playground) - great for rigid pattern/structural matching, which may work well for issue analysis

**Comment by modichika** [comment]
Hey @alyssacgoins the problem with gh models are that github has set 4000 or 8000 tokens on every model by default so even if I changed the model and researched on loads of them this issue https://github.com/modichika/pipelines/issues/72
gave same limit error, I have solved the crashing error it won't crash even if the rate limit is there.

One concern is there: this pre-approved issue - https://github.com/kubeflow/pipelines/issues/13180  was created by @droctothorpe(a maintainer) do we need llm to analyse these kinds of issues?

Please look at the above test maybe we can come up with some solutions what are your thought?

File: .github/scripts/ai_issue_triage.sh [hunk 6]
```diff
@@ -0,0 +1,56 @@
+set -e
+gh extension install github/gh-models --force || true
+
+echo "Using model: $TARGET_MODEL"
+
+read -r -d '' SYSTEM_INSTRUCTIONS << 'EOF'  || true
+You are an expert open-source maintainer for Kubeflow.
```
Comment (zazulam) [inline .github/scripts/ai_issue_triage.sh]: Let's be specific and say `Kubeflow Pipelines`

File: .github/scripts/ai_issue_triage.sh [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+- BUG TIER (#13180): [bug] fix: E2E test flakiness on K8s v1.34.0 — root cause analysis. High-quality root-cause analysis and environment data.
+- FRONTEND (#13108): [frontend] Adds coverage for frontend mock:api startup and enum drift. Explicit file paths and definitions of done.
+- SDK (#12865): [sdk] [bug] [set_accelerator_limit] rejects valid accelerator counts (only allows 0, 1, 2, 4, 8, 16). Elite precision with failing parameters.
+
+Compare the incoming issue detail density directly against the relevant blueprint standard above.
+
+Respond strictly following this format structure without other markdown wraps:
+
+### 📊 Scope
+<Evaluate task boundaries and estimate the time window required to complete it>
```
Comment (zazulam) [inline .github/scripts/ai_issue_triage.sh]: Let's remove time window as it does not aid much in communicating the bug/feat scope.

File: .github/scripts/ai_issue_triage.sh [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+
+### 📝 Context & Guidance
+<Evaluate if steps, expected behavior, or links are provided against repo standards>
+
+### ⚡ Complexity
+<Rate difficulty: Low, Medium, High, detailing architectural deepness vs shallow tweaks>
```
Comment (zazulam) [inline .github/scripts/ai_issue_triage.sh]: If we don't define the the difficulty levels then the llm will determine it. There should be a mapping of some kind that where the breadth is taken into consideration along with the depth i.e. a bug fix that spans the sdk and argo compiler vs a feature that touches the sdk, frontend, backend that adds new modules in each.

File: .github/scripts/ai_triage.sh [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+EOF
+fi
+
+echo "analysis<<EOF" >> $GITHUB_OUTPUT
+echo "$ANALYSIS_REPORT" >> $GITHUB_OUTPUT
+echo "EOF" >> $GITHUB_OUTPUT
+
+          
+echo "DEBUG: The raw analysis sent to output was:"
+echo "$ANALYSIS_REPORT"
```
Comment (alyssacgoins) [inline .github/scripts/ai_triage.sh]: EOF

File: .github/workflows/ai_labeler.yml [hunk 10]
```diff
@@ -0,0 +1,37 @@
...
+          ./.github/scripts/ai_triage.sh
+
+      - name: Apply Quality Analysis Comment
+        env:
+          GH_TOKEN: ${{ secrets.GITHUB_TOKEN }}
+          ISSUE_NUMBER: ${{ github.event.issue.number }}
+          REPORT: ${{ steps.ai_logic.outputs.analysis }}
+        run: |
+          gh issue comment "$ISSUE_NUMBER" --body "## 🤖 AI Issue Quality Review
+          $REPORT"
```
Comment (alyssacgoins) [inline .github/workflows/ai_labeler.yml]: EOF

File: .github/scripts/ai_triage.sh [hunk 11]
```diff
@@ -0,0 +1,56 @@
...
+             
+Calibrate your evaluation against these compressed reference standards:
+- BACKEND (#13314): [backend] S3 operations fail with non-AWS object stores after AWS SDK v2 checksum defaults change. Clear isolated scope.
+- BUG TIER (#13180): [bug] fix: E2E test flakiness on K8s v1.34.0 — root cause analysis. High-quality root-cause analysis and environment data.
+- FRONTEND (#13108): [frontend] Adds coverage for frontend mock:api startup and enum drift. Explicit file paths and definitions of done.
+- SDK (#12865): [sdk] [bug] [set_accelerator_limit] rejects valid accelerator counts (only allows 0, 1, 2, 4, 8, 16). Elite precision with failing parameters.
+
+Compare the incoming issue detail density directly against the relevant blueprint standard above.
+
+Respond strictly following this format structure without other markdown wraps:
```
Comment (alyssacgoins) [inline .github/scripts/ai_triage.sh:19]: I like that the the AI review is divided into sections - makes it very readable. I think we could potentially shorten the length of the review, however - I think you could cut down the size or switch the paragraphs to bullet points, for maintainers/approvers looking to quickly understand the issue analysis

File: .github/scripts/ai_triage.sh (hunk 9 above)
Comment (modichika) [reply .github/scripts/ai_triage.sh]: hey @alyssacgoins I didn't understand this can you please elaborate?

Comment (alyssacgoins) [reply .github/scripts/ai_triage.sh]: @modichika EOF means end of file - the red circle above appears when there is a newline missing at the end of the file - can you add it?

File: .github/workflows/ai_labeler.yml (hunk 10 above)
Comment (alyssacgoins) [reply .github/workflows/ai_labeler.yml]: @modichika same as above

**Review by alyssacgoins — CHANGES_REQUESTED** [review]

File: .github/scripts/ai_triage.sh (hunk 9 above)
Comment (modichika) [reply .github/scripts/ai_triage.sh]: yeah added it!

File: .github/scripts/ai_triage.sh [hunk 12]
```diff
@@ -0,0 +1,69 @@
+set -e
+gh extension install github/gh-models --force || true
+
+echo "Using model: $TARGET_MODEL"
+
+read -r -d '' SYSTEM_INSTRUCTIONS << 'EOF'  || true
+You are an expert open-source maintainer for Kubeflow pipelines.
```
Comment (mprahl) [inline .github/scripts/ai_triage.sh]:
```suggestion
You are an expert open-source maintainer for Kubeflow Pipelines.
```

File: .github/workflows/ai_labeler.yml [hunk 13]
```diff
@@ -0,0 +1,38 @@
...
+        uses: actions/checkout@v4
+
+      - name: Execute AI Issue Triage
+        id: ai_logic
+        env:
+          GH_TOKEN: ${{ secrets.GITHUB_TOKEN }}
+          TITLE: ${{ github.event.issue.title }}
+          RAW_BODY: ${{ github.event.issue.body }}
+        run: |
+          chmod +x ./.github/scripts/ai_triage.sh
```
Comment (mprahl) [inline .github/workflows/ai_labeler.yml]: Nit: I think we can just commit the script with `chmod +x` already applied.

File: .github/workflows/ai_analyzer.yml [hunk 14]
```diff
@@ -0,0 +1,37 @@
...
+          ./.github/scripts/ai_triage.sh
+
+      - name: Apply Quality Analysis Comment
+        env:
+          GH_TOKEN: ${{ secrets.GITHUB_TOKEN }}
+          ISSUE_NUMBER: ${{ github.event.issue.number }}
+          REPORT: ${{ steps.ai_logic.outputs.analysis }}
+        run: |
+          gh issue comment "$ISSUE_NUMBER" --body "## 🤖 AI Issue Quality Review
+          $REPORT"
```
Comment (mprahl) [inline .github/workflows/ai_analyzer.yml:36]: The workflow never applies any labels. This step only creates a comment, and the earlier model step only emits prose, so the `AI Issue Labeler` job will not perform its advertised labeling behavior. Please add a label-selection output and an API/gh step that applies allowed repository labels, or rename/scope the workflow to issue-quality commenting.

File: .github/scripts/ai_triage.sh [hunk 15]
```diff
@@ -0,0 +1,69 @@
...
+Compare the incoming issue detail density directly against the relevant blueprint standard above.
+
+Respond strictly following this format structure without other markdown wraps:
+
+- Each section MUST contain exactly 2 to 3 short, bullet fragments. 
+- Do NOT write full-length paragraphs or introductory text. Keep it highly concise for quick scanning.
+- Do NOT include any time frame or implementation window estimations.
+
+
+### 📊 Scope
```
Comment (mprahl) [inline .github/scripts/ai_triage.sh:26]: The system prompt does not ask the model to select labels or constrain it to this repo's label taxonomy. Before using model output for labeling, please include an explicit allowlist, selection rules, and a strict machine-readable schema such as JSON with `labels` and `comment`; otherwise the workflow will be forced to parse free-form prose and may request labels that do not exist.

File: .github/workflows/ai_labeler.yml [hunk 16]
```diff
@@ -0,0 +1,37 @@
+name: AI Issue Labeler
+on:
+  issues:
+    types: [opened, edited]
```
Comment (mprahl) [inline .github/workflows/ai_labeler.yml]: Because this also runs on every issue edit and always calls `gh issue comment`, routine edits can create repeated AI comments and model calls on the same issue. If edited-trigger retriage is needed, please update an existing bot comment or add gating/rate limiting so the workflow does not spam issue threads.

**Review by mprahl — COMMENTED** [review]
PR Overview

This adds an AI triage workflow and shell script, but it currently comments on issues rather than labeling them.

File: .github/workflows/ai_analyzer.yml (hunk 14 above)
Comment (modichika) [reply .github/workflows/ai_analyzer.yml:36]: yeah renamed the workflow!

File: .github/scripts/ai_triage.sh (hunk 15 above)
Comment (modichika) [reply .github/scripts/ai_triage.sh:26]: now it's to add comment for maintainers and not to add tags.

**Comment by modichika** [comment]
Thanks for the review @mprahl 
updated the workflow name, inner definitions from 'Labeler' to 'Analyzer' inside the code to  its behavior (commenting analysis prose). 
the trigger responds to opened to avoid many comments on edits

**Review by mprahl — APPROVED** [review]
/approve
/lgtm
